brainstorming.md

```diff
@@ -0,0 +1,372 @@
+
+Where SOMA requires an explicit typing system, it utilizes the Arrow types and schema. SOMA has no specific requirements for the type or serialization system used by the underlying storage engine, other than it be capable of understanding and representing the Arrow types. It is understood and expected that each implementation of SOMA will have data type expressiveness limits (eg, just because you can express a type in he Arrow type system does not mean all SOMA implementations will understand it).
+
+In the following doc:
+
+- `primitive` types in this specification refers to Arrow primitive types, eg, `int32`, `float`, etc.
+- `string` refers to Arrow UTF-8 variable-length `string`, ie, `List<Char>`.
+- `simple` types include all primitive types, plus `string`.
+
+Other Arrow types are explicitly noted as such, eg, `Arrow.RecordBatch`.
+
+> ⚠️ **Issue** - is the use of Arrow as a base type language acceptable? There are others, up to and including inventing a new system, but it seemed prudent to adopt an existing, language neutral specification, which is already supported by many "engines".
+
+> ⚠️ **Issue** - are there parts of the Arrow type system that we wish to _explicitly exclude_ from SOMA? I have left this issue open (ie, no specific text) for now, thinking that we can come back and subset as we understand what complex types are required, and how much flexibility should be in this spec. We clearly need some complex types (eg, RecordBatch, List, etc) as they are implied by `string`, etc. My own preference would be to mandate a small set of primitive types, and leave the rest open to implementations to support as they feel useful.
+
+## Metadata (uns)
+
+All SOMA objects may be annotated with a small amount of metadata, in the form of a simple key/value map:
+
+- `uns` (map[string, simple]) - string-keyed mapping to any **simple** type (eg, int32, string, etc).
+
+Only `simple` types are supported in the metadata values. The metadata lifecycle is the same as its containing object, eg, it will be deleted when the containing object is deleted.
+
+> ⚠️ **Issue** - `uns` - do we need more complex typing for uns values, and if so, how to represent capture what can is required to be supported? (eg, anything Arrow can represent? a subset?)
+>
+> **Proposal**: restrict `uns` to a mapping of string keys and `simple` values. Extend to complex types in the future as we have concrete use cases which warrant the complexity.
+
+> ⚠️ **Issue** - the name `uns` is not really all that suggestive. **Proposal:** rename `uns` to `metadata`, eg, `print(soma.metadata['version'])`
+
+## SOMACollection
+
+SOMACollection is an unordered, `string`-keyed map of values of type SOMADataFrame, SOMAMatrix, SOMACollection or SOMA. SOMACollection objects are typically persistent, and may be nested to any level (eg, SOMACollection may contain other SOMACollection objects). Keys in the map are unique and singular (no duplicates, ie, it is not a multi-map).
```

Comment:
theoretically to any level but realistically when would we ever want to go beyond three? corpus -> multimodal -> each mode

Reply:
does allowing (encouraging) arbitrary depth help us?

Reply:
... i just read below & now see you use `SOMACollection` for things like containing the `obsm` element arrays so i retract this now ;)

Reply:
... what i am now (having read through) thinking of as the 'outer role' and the 'inner role' which `SOMACollection plays both of

Reply:
you have the essence of this right (or at least as I intended it).  Both use cases are persistent, and the flexibility to nest will be important for some use cases (eg, "atlas" of "multi-modal", etc).

I'm adding a bit of expository prose as you suggested to make this clearer.

Reply:
@johnkerl - please resolve this thread if you are OK with the changes to address this concern.